src/data/matrix/reflection.lean

```diff
@@ -0,0 +1,481 @@
+  let a_flat := (list.fin_range l).bind (λ i, (list.fin_range m).map $ λ j, a i j),
+  let b_flat := (list.fin_range m).bind (λ i, (list.fin_range n).map $ λ j, b i j),
+  let args := [α, has_mul_α, add_comm_monoid_α] ++ a_flat ++ b_flat,
+
+  -- build the matrices out of the coefficients
+  let A := matrix.fin_to_pexpr (matrix.map a to_pexpr),
+  let B := matrix.fin_to_pexpr (matrix.map b to_pexpr),
+  -- get an instance cache holding all the instances needed for matrix multiplication. There must
+  -- be a better way to do this.
+  t ← tactic.mk_instance_cache α,
+  has_add_α ← tactic.mk_app `has_add [α] >>= (λ t, prod.snd <$> @tactic.solve_aux unit t (do
+  { tmp2 ← tactic.pose `_inst_2' none add_comm_monoid_α,
+    tactic.reset_instance_cache,
+    `[apply_instance] })),
+  has_zero_α ← tactic.mk_app `has_zero [α] >>= (λ t, prod.snd <$> @tactic.solve_aux unit t (do
+  { tmp2 ← tactic.pose `_inst_2' none add_comm_monoid_α,
+    tactic.reset_instance_cache,
+    `[apply_instance] })),
+  let t := {inst := [
+    (`has_mul, has_mul_α),
+    (`has_add, has_add_α),
+    (`has_zero, has_zero_α),
+    (`add_comm_monoid, add_comm_monoid_α)].foldl (λ n x, n.insert x.1 x.2) t.inst,
+     ..t},
+
+  -- clever trick: create algebraic instances on `expr` so that we can use `matrix.mul` or
+  -- `matrix.mulᵣ` to build the expression we want to end up with. It doesn't matter which we pick,
+  -- but the typeclasses are easier to create for the latter.
+  (t, has_mul_αe) ← expr.has_mul t,
+  (t, has_add_αe) ← expr.has_add t,
+  (t, has_zero_αe) ← expr.has_zero t,
+  let ab := @matrix.mulᵣ _ _ _ _ has_mul_αe has_add_αe has_zero_αe a b,
+  let AB := matrix.fin_to_pexpr (matrix.map ab to_pexpr),
+
+  -- State and prove the equality, noting the RHS is defeq to `mulᵣ A B`.
+  A_eq ← tactic.to_expr ``(@matrix.mul _ _ _ _ _ %%has_mul_α %%add_comm_monoid_α %%A %%B = %%AB),
+  t ← tactic.pis args A_eq,
+  let pr := (expr.const `matrix.mul_fin_aux [u]).mk_app [`(l), `(m), `(n)],
+  -- This seems to create a metavariable then assign it, which ensures `pr` carries the right type.
+  ((), pr) ← tactic.solve_aux t $ tactic.exact pr,
+
+  pure (t, pr)
+
+open_locale matrix
+
+
+/-- Helper tactic used as an `auto_param` for `matrix.mul_fin` -/
+meta def mul_fin.derive : tactic unit :=
+do
+  target@`(@matrix.mul (fin %%l) (fin %%m) (fin %%n) %%α %%_ %%i1 %%i2 %%A %%B = %%AB)
+    ← tactic.target,
+  some (l, m, n) ← pure (prod.mk <$> l.to_nat <*> (prod.mk <$> m.to_nat <*> n.to_nat)) |
+    fail!"Dimensions {l}, {m} {n} are not numerals",
+  (t,pr) ← mul_fin.prove l m n,
+  tactic.apply (pr α i1 i2) {},
+  tactic.done
+  -- TODO: should we be extracting the coefficients manually so we can do a full invocation as
+  -- something like:
+  --   tactic.unify target (t.instantiate_pis [α, A']),
+  --   tactic.exact (pr α A')
+
+
+theorem mul_fin {α} [has_mul α] [add_comm_monoid α] {l m n : ℕ}
+  {«![a₁₁ ⋱ aₗₘ]» : fin l → fin m → α}
```

Comment:
These weird variable names look like some kind of magic, but they actually aren't.
How about just using normal variable names like `A` and `B` and `AB` and, if necessary, explaining the intended use (or whatever else the names were supposed to convey) in a comment?

Reply:
The goal here is for `#check mul_fin` to print out a reasonable summary of what the shape of the generated lemma will look like; in this case, the lemma is only intended to be used with `of ![![a]] ⬝ of ![![b]]  = of ![![a * b]]` and similar, where `«![a₁₁ ⋱ aₗₘ]» =  ![![a]]` is deduced by unification with the auto_param.

Reply:
`#print mul_fin` would be better anyways, and display the docstring.

Reply:
Oh, I'm totally happy to explain what's going on here in a docstring; but I think changing this to `A` and `B` would obscure the fact that something magic _is_ happening somewhere here.

Reply:
FWIW I think I agree with @rwbarton here.

I find the magic-looking but not-actually-magic variable names hard to read and I think it's reasonable for the doc string to take on the job of explanation.

Reply:
I've renamed the arguments for this one. I wasn't sure about `matrix.fin_eta`.